README.md
# Http Downloader

Simply download your files from the internet and save them to the desired location on your device.

Download your files from the internet and see the progress while downloading. Edit the progress bar as you wish. You can adjust the view before the download starts and after the download is complete. If you do not set it, it will remain as default.

## Usage

1. Use `HttpDownloaderWidget()` where the download progress will be.
3. HttpDownloaderWidget(<br>
`url:` (Link of the file you want to download `ex. https://google.com/example.png`),<br>
`path:` (Location to save the file `ex. C:\\Users\\YourName\\Desktop`),<br>
`fileName:` (The name to save the file `ex. image.png`)<br>
):

## Example

If you do this example, it will have the default view.
```
HttpDownloaderWidget(
  url: "https://google.com/example.png",
  path: "C:\\Users\\YourName\\Desktop",
  fileName: "image.png"
)
```

You can customize it this way.
```
HttpDownloaderWidget(
  url: "https://google.com/example.png",
  path: "C:\\Users\\YourName\\Desktop",
  fileName: "image.png"
  startDownloadWidget: (download) {
    return ElevatedButton(
      onPressed: download,
      child: Text("Download"),
    );
  },
  downloadingWidget: (progress, downloadedMB, contentMB) {
    return Column(
      children: [
        Text("Downloading.. ${progress.toString()}%", style: TextStyle(color: Colors.black)),
        SizedBox(
          width: 400,
          child: LinearProgressIndicator(
            value: progress / 100,
            borderRadius: BorderRadius.circular(10),
            color: Colors.red,
          ),
        ),
        Text("$downloadedMB MB / $contentMB MB", style: TextStyle(color: Colors.amber)),
      ],
    );
  },
  downloadedWidget: const Text("Download Completed"),
)
```


CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* Added customizable downloader and Readme


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:http_downloader/http_downloader_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
      ),
      body: const Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            HttpDownloaderWidget(
              url: "URL",
              path: "PATH",
              fileName: "FILENAME",
            ),
          ],
        ),
      ),
    );
  }
}


lib/http_downloader.dart
library http_downloader;

import 'dart:async';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

typedef DownloadProgress = void Function(
    int total, int downloaded, double progress);

class HttpDownloader {
  static Future<Uint8List> download(
      String url, DownloadProgress downloadProgress) async {
    final compelter = Completer<Uint8List>();
    final client = http.Client();
    final request = http.Request('GET', Uri.parse(url));
    final response = client.send(request);

    int downloadedBytes = 0;
    List<List<int>> chunkList = [];

    response.asStream().listen((http.StreamedResponse streamedResponse) {
      streamedResponse.stream.listen(
        (chunk) {
          final contentLength = streamedResponse.contentLength ?? 0;
          final progress = (downloadedBytes / contentLength) * 100;
          downloadProgress(contentLength, downloadedBytes, progress);

          chunkList.add(chunk);
          downloadedBytes += chunk.length;
        },
        onDone: () {
          final contentLength = streamedResponse.contentLength ?? 0;
          final progress = (downloadedBytes / contentLength) * 100;
          downloadProgress(contentLength, downloadedBytes, progress);

          int start = 0;
          final bytes = Uint8List(contentLength);

          for (var chunk in chunkList) {
            bytes.setRange(start, start + chunk.length, chunk);
            start += chunk.length;
          }

          compelter.complete(bytes);
        },
        onError: (error) => compelter.completeError(error),
      );
    });

    return compelter.future;
  }
}


lib/downloader_widget.dart
library downloader_widget;

import 'dart:developer';
import 'dart:io';
import 'dart:typed_data';
import 'package:flutter/material.dart';
import 'package:http_downloader/http_downloader.dart';

class DownloaderWidget extends StatefulWidget {
  const DownloaderWidget({super.key, required this.url, required this.path});
  final String url;
  final String path;
  @override
  State<DownloaderWidget> createState() => _DownloaderWidgetState();
}

class _DownloaderWidgetState extends State<DownloaderWidget> {
  bool downloading = false;
  double dlProgress = 0;
  int contentLength = 0;
  int downloadedBytes = 0;
  Uint8List? uint8list;
  bool shortcutChecked = true;
  bool downloaded = false;

  Future<void> donwloadFile() async {
    downloading = true;
    setState(() {});
    try {
      final bytes = await HttpDownloader.download(widget.url, (total, downloaded, progress) {
        contentLength = total;
        downloadedBytes = downloaded;
        dlProgress = progress;

        setState(() {});
      });
      try {
        File(widget.path)
          ..createSync(recursive: true)
          ..writeAsBytesSync(bytes);
        print('File Downloaded: ${widget.path}');
      } catch (e) {
        print("Error: $e");
      }
    } catch (e) {
      log('$e');
    }
    downloading = false;
    downloaded = true;
    setState(() {});
  }

  Future<String> checkDownload() async {
    if (downloaded) {
      return "downloaded";
    } else {
      if (downloading) {
        return "downloading";
      } else {
        donwloadFile();
        return "download";
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: checkDownload(),
      builder: (_, snapshot) {
        if (snapshot.data == "downloading") {
          return Column(
            children: [
              Text("Downloading... ${dlProgress.toInt().toString()}%", style: TextStyle(color: Colors.black)),
              SizedBox(
                width: 400,
                child: LinearProgressIndicator(
                  value: dlProgress / 100,
                  borderRadius: BorderRadius.circular(10),
                  color: Colors.red,
                ),
              ),
              Text("${(downloadedBytes / 1048576).toStringAsFixed(2)} MB / ${(contentLength / 1048576).toStringAsFixed(2)} MB", style: TextStyle(color: Colors.black)),
            ],
          );
        } else if (snapshot.data == "downloaded") {
          return Text("Downloaded");
        } else {
          return Container();
        }
      },
    );
  }
}


lib/http_downloader_widget.dart
library http_downloader_widget;

import 'dart:developer';
import 'dart:io';
import 'dart:typed_data';
import 'package:flutter/material.dart';
import 'package:http_downloader/http_downloader.dart';

class HttpDownloaderWidget extends StatefulWidget {
  const HttpDownloaderWidget({
    super.key,
    required this.url,
    required this.path,
    required this.fileName,
    this.startDownloadWidget,
    this.downloadingWidget,
    this.downloadedWidget,
  });
  final String url;
  final String path;
  final String fileName;
  final Widget Function(Future<void> Function() download)? startDownloadWidget;
  final Widget Function(int progress, String downloadedMB, String contentMB)?
      downloadingWidget;
  final Widget? downloadedWidget;
  @override
  State<HttpDownloaderWidget> createState() => _HttpDownloaderWidgetState();
}

class _HttpDownloaderWidgetState extends State<HttpDownloaderWidget> {
  bool downloading = false;
  double dlProgress = 0;
  int contentLength = 0;
  int downloadedBytes = 0;
  Uint8List? uint8list;
  bool shortcutChecked = true;
  bool downloaded = false;

  Future<void> downloadFile() async {
    downloading = true;
    setState(() {});
    try {
      final bytes = await HttpDownloader.download(widget.url,
          (total, downloaded, progress) {
        contentLength = total;
        downloadedBytes = downloaded;
        dlProgress = progress;

        setState(() {});
      });
      try {
        File("${widget.path}\\${widget.fileName}")
          ..createSync(recursive: true)
          ..writeAsBytesSync(bytes);
        log('File Downloaded: ${widget.path}\\${widget.fileName}');
      } catch (e) {
        log("Error: $e");
      }
    } catch (e) {
      log('$e');
    }
    downloading = false;
    downloaded = true;
    setState(() {});
  }

  Future<String> checkDownload() async {
    if (downloaded) {
      return "downloaded";
    } else {
      if (downloading) {
        return "downloading";
      } else {
        return "download";
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: checkDownload(),
      builder: (_, snapshot) {
        if (snapshot.data == "downloading") {
          return widget.downloadingWidget == null
              ? Column(
                  children: [
                    Text("Downloading... ${dlProgress.toInt().toString()}%",
                        style: const TextStyle(color: Colors.black)),
                    SizedBox(
                      width: 400,
                      child: LinearProgressIndicator(
                        value: dlProgress / 100,
                        borderRadius: BorderRadius.circular(10),
                        color: Colors.red,
                      ),
                    ),
                    Text(
                        "${(downloadedBytes / 1048576).toStringAsFixed(2)} MB / ${(contentLength / 1048576).toStringAsFixed(2)} MB",
                        style: const TextStyle(color: Colors.black)),
                  ],
                )
              : widget.downloadingWidget!(
                  dlProgress.toInt(),
                  (downloadedBytes / 1048576).toStringAsFixed(2),
                  (contentLength / 1048576).toStringAsFixed(2));
        } else if (snapshot.data == "download") {
          return widget.startDownloadWidget == null
              ? ElevatedButton(
                  onPressed: () {
                    downloadFile();
                  },
                  child: const Text("Start Download"),
                )
              : widget.startDownloadWidget!(downloadFile);
        } else if (snapshot.data == "downloaded") {
          return widget.downloadedWidget ?? const Text("Downloaded");
        } else {
          return Container();
        }
      },
    );
  }
}
